src/feature/dircache/dircache.c

```diff
@@ -495,28 +495,49 @@ handle_get_frontpage(dir_connection_t *conn, const get_handler_args_t *args)
 }
 
 /** Warn that the cached consensus <b>consensus</b> of type
- * <b>flavor</b> is too old and will not be served to clients. Rate-limit the
- * warning to avoid logging an entry on every request.
+ * <b>flavor</b> too new or too old, based on <b>is_too_new</b>,
+ * and will not be served to clients. Rate-limit the warning to avoid logging
+ * an entry on every request.
  */
 static void
-warn_consensus_is_too_old(const struct consensus_cache_entry_t *consensus,
-                          const char *flavor, time_t now)
+warn_consensus_is_not_reasonably_live(
+                          const struct consensus_cache_entry_t *consensus,
+                          const char *flavor, time_t now, bool is_too_new)
 {
-#define TOO_OLD_WARNING_INTERVAL (60*60)
-  static ratelim_t warned = RATELIM_INIT(TOO_OLD_WARNING_INTERVAL);
+#define NOT_REASONABLY_LIVE_WARNING_INTERVAL (60*60)
+  static ratelim_t warned[2] = { RATELIM_INIT(
+                                      NOT_REASONABLY_LIVE_WARNING_INTERVAL),
+                                RATELIM_INIT(
+                                      NOT_REASONABLY_LIVE_WARNING_INTERVAL) };
   char timestamp[ISO_TIME_LEN+1];
-  time_t valid_until;
+  /* valid_after if is_too_new, valid_until if !is_too_new */
+  time_t valid_time;
   char *dupes;
 
-  if (consensus_cache_entry_get_valid_until(consensus, &valid_until))
+  if (is_too_new) {
+    if (consensus_cache_entry_get_valid_after(consensus, &valid_time))
+      return;
+  } else {
+    if (consensus_cache_entry_get_valid_until(consensus, &valid_time))
+      return;
+  }
+
+  /* Check that is_too_new is a valid array index */
+  if (BUG(is_too_new != 0 && is_too_new != 1))
```

Comment:
It looks like gcc ((Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609) is clever enough to know that bools only have two values.

```
In file included from ../src/core/or/or.h:26:0,
                 from ../src/feature/dircache/dircache.c:8:
../src/feature/dircache/dircache.c: In function ‘warn_consensus_is_not_reasonably_live’:
../src/feature/dircache/dircache.c:526:27: error: logical ‘and’ of mutually exclusive tests is always false [-Werror=logical-op]
   if (BUG(is_too_new != 0 && is_too_new != 1))
                           ^
../src/lib/cc/compat_compiler.h:149:51: note: in definition of macro ‘PREDICT_UNLIKELY’
 #define PREDICT_UNLIKELY(exp) __builtin_expect(!!(exp), 0)
                                                   ^
../src/lib/log/util_bug.h:178:4: note: in expansion of macro ‘ASSERT_PREDICT_UNLIKELY_’
   (ASSERT_PREDICT_UNLIKELY_(cond) ?                                     \
    ^
../src/feature/dircache/dircache.c:526:7: note: in expansion of macro ‘BUG’
   if (BUG(is_too_new != 0 && is_too_new != 1))
       ^
cc1: all warnings being treated as errors
```